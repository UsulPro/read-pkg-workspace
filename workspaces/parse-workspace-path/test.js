const path = require('path');
const chalk = require('chalk');
const parseWspPath = require('./index');


const rootPath = path.resolve(__dirname, '../');
const resultsToColorize = [];

const colorize = result => {
  const output = [
    result.dir,
    result.workspaceName ? `[${chalk.cyan(result.workspaceName)}]` : '',
    chalk.greenBright(result.packageDir),
    result.packageName ? chalk.yellow(`(${result.packageName}@${result.packageVersion})`) : '',
    chalk.white(result.localDir),
    chalk.magenta(result.base),
  ].filter(e => e).join(path.sep);
  return chalk.green(output);
}

const testSnapshot = result => {
  const mockedDir = {
    ...result,
    dir: 'dir',
  }
  expect(mockedDir).toMatchSnapshot();
}

const testPathSequence = (cwd, result) => {
  // console.log('CWD: ', cwd, '\n\n', result);
  const summaryPath = path.resolve(
    result.dir,
    result.workspaceName,
    result.packageDir,
    result.localDir,
    result.base,
  );
  expect(summaryPath).toBe(cwd);
  resultsToColorize.push(result);
  testSnapshot(result);
}

describe('parse-workspace-path. use cases', () => {
  it('test package in the workspace (direclty specified)', () => {
    const cwd = path.resolve(rootPath, 'read-pkg-workspace');

    expect.hasAssertions();
    return parseWspPath(cwd)
      .then(result => {
        testPathSequence(cwd, result);
      })
  });

  it('test package in the workspace (specified by *)', () => {
    const cwd = path.resolve(rootPath, 'unicorns/celestabelleabethabelle/src/stories/');

    expect.hasAssertions();
    return parseWspPath(cwd)
      .then(result => {
        testPathSequence(cwd, result);
      })
  });

  it('test file in workspace', () => {
    const cwd = path.resolve(rootPath, 'unicorns/celestabelleabethabelle/src/stories/unicorn.story.js');

    expect.hasAssertions();
    return parseWspPath(cwd)
      .then(result => {
        testPathSequence(cwd, result);
      })
  });

  it('test file in a single package', () => {
    const cwd = path.resolve(rootPath, '../usual-packages/recreation-space/src/index.js');

    expect.hasAssertions();
    return parseWspPath(cwd)
      .then(result => {
        testPathSequence(cwd, result);
      })
  });

  it('test file in a package not specified in workspaces', () => {
    const cwd = path.resolve(rootPath, 'usual-packages/recreation-space/src/index.js');

    expect.hasAssertions();
    return parseWspPath(cwd)
      .then(result => {
        testPathSequence(cwd, result);
      })
  });

  it('test nested packages dir', () => {
    const cwd = path.resolve(rootPath, 'usual-packages/recreation-space/node_modules/read-pkg/dist');

    expect.hasAssertions();
    return parseWspPath(cwd)
      .then(result => {
        testPathSequence(cwd, result);
      })
  });

  it('test nested packages file', () => {
    const cwd = path.resolve(rootPath, 'usual-packages/recreation-space/node_modules/read-pkg/index.js');

    expect.hasAssertions();
    return parseWspPath(cwd)
      .then(result => {
        testPathSequence(cwd, result);
      })
  });

  it('test single workspace dir', () => {
    const cwd = path.resolve(rootPath, 'usual-packages/');

    expect.hasAssertions();
    return parseWspPath(cwd)
      .then(result => {
        testPathSequence(cwd, result);
      })
  });

  it('test single workspace file', () => {
    const cwd = path.resolve(rootPath, 'usual-packages/readme.md');

    expect.hasAssertions();
    return parseWspPath(cwd)
      .then(result => {
        testPathSequence(cwd, result);
      })
  });

  it('test no package dir', () => {
    const cwd = path.resolve(rootPath, '../');

    expect.hasAssertions();
    return parseWspPath(cwd)
      .then(result => {
        testPathSequence(cwd, result);
      })
  });

  it('test no package file', () => {
    const cwd = path.resolve(rootPath, '../readme.md');

    expect.hasAssertions();
    return parseWspPath(cwd)
      .then(result => {
        testPathSequence(cwd, result);
      })
  });

  it('test nonexistent file', () => {
    const cwd = path.resolve(rootPath, 'unicorns/celestabelleabethabelle/src/stories/unicorn.story.jsx');

    expect.hasAssertions();
    return parseWspPath(cwd)
      .catch(err => {
        expect(err).toHaveProperty('code', 'ENOENT');
      })
  });

})

describe('parse-workspace-path. color output', () => {
  it('visual tests:', () => {
    expect.assertions(11);
    process.stdout.write([
      chalk.gray('\nLegend:\t'),
      chalk.green('dir/['),
      chalk.cyan('workspace'),
      chalk.green(']/'),
      chalk.greenBright('packageDir/'),
      chalk.yellow('(packageName@packageVersion)'),
      chalk.white('/localDir/'),
      chalk.magenta('base'),
      '\n\n',
    ].join(''))
    resultsToColorize.forEach(result => {
      expect(() => process.stdout.write(`\t${colorize(result)}\n`)).not.toThrow();
    })
    process.stdout.write(`\n\n`);

  });
})

