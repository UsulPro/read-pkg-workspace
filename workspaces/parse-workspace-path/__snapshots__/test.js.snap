// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse-workspace-path. use cases test file in a package not specified in workspaces 1`] = `
Object {
  "base": "index.js",
  "dir": "dir",
  "ext": ".js",
  "localDir": "src",
  "name": "index",
  "packageDir": "",
  "packageName": "recreation-space",
  "packageVersion": "1.0.0",
  "root": "/",
  "workspaceName": "",
}
`;

exports[`parse-workspace-path. use cases test file in a single package 1`] = `
Object {
  "base": "index.js",
  "dir": "dir",
  "ext": ".js",
  "localDir": "src",
  "name": "index",
  "packageDir": "",
  "packageName": "recreation-space",
  "packageVersion": "1.0.0",
  "root": "/",
  "workspaceName": "",
}
`;

exports[`parse-workspace-path. use cases test file in workspace 1`] = `
Object {
  "base": "unicorn.story.js",
  "dir": "dir",
  "ext": ".js",
  "localDir": "src/stories",
  "name": "unicorn.story",
  "packageDir": "unicorns/celestabelleabethabelle",
  "packageName": "celestabelleabethabelle",
  "packageVersion": "1.0.0",
  "root": "/",
  "workspaceName": "workspaces",
}
`;

exports[`parse-workspace-path. use cases test nested packages dir 1`] = `
Object {
  "base": "dist",
  "dir": "dir",
  "ext": "",
  "localDir": "",
  "name": "dist",
  "packageDir": "",
  "packageName": "read-pkg",
  "packageVersion": "3.0.0",
  "root": "/",
  "workspaceName": "",
}
`;

exports[`parse-workspace-path. use cases test nested packages file 1`] = `
Object {
  "base": "index.js",
  "dir": "dir",
  "ext": ".js",
  "localDir": "",
  "name": "index",
  "packageDir": "",
  "packageName": "read-pkg",
  "packageVersion": "3.0.0",
  "root": "/",
  "workspaceName": "",
}
`;

exports[`parse-workspace-path. use cases test no package dir 1`] = `
Object {
  "base": "",
  "dir": "dir",
  "ext": "",
  "localDir": "",
  "name": "",
  "packageDir": "",
  "packageName": "",
  "packageVersion": "",
  "root": "/",
  "workspaceName": "",
}
`;

exports[`parse-workspace-path. use cases test no package file 1`] = `
Object {
  "base": "",
  "dir": "dir",
  "ext": "",
  "localDir": "",
  "name": "",
  "packageDir": "",
  "packageName": "",
  "packageVersion": "",
  "root": "/",
  "workspaceName": "",
}
`;

exports[`parse-workspace-path. use cases test package in the workspace (direclty specified) 1`] = `
Object {
  "base": "",
  "dir": "dir",
  "ext": "",
  "localDir": "",
  "name": "",
  "packageDir": "read-pkg-workspace",
  "packageName": "read-pkg-workspace",
  "packageVersion": "0.1.0",
  "root": "/",
  "workspaceName": "workspaces",
}
`;

exports[`parse-workspace-path. use cases test package in the workspace (specified by *) 1`] = `
Object {
  "base": "stories",
  "dir": "dir",
  "ext": "",
  "localDir": "src",
  "name": "stories",
  "packageDir": "unicorns/celestabelleabethabelle",
  "packageName": "celestabelleabethabelle",
  "packageVersion": "1.0.0",
  "root": "/",
  "workspaceName": "workspaces",
}
`;

exports[`parse-workspace-path. use cases test single workspace dir 1`] = `
Object {
  "base": "usual-packages",
  "dir": "dir",
  "ext": "",
  "localDir": "",
  "name": "usual-packages",
  "packageDir": "",
  "packageName": "",
  "packageVersion": "",
  "root": "/",
  "workspaceName": "workspaces",
}
`;

exports[`parse-workspace-path. use cases test single workspace file 1`] = `
Object {
  "base": "readme.md",
  "dir": "dir",
  "ext": ".md",
  "localDir": "usual-packages",
  "name": "readme",
  "packageDir": "",
  "packageName": "",
  "packageVersion": "",
  "root": "/",
  "workspaceName": "workspaces",
}
`;
